ruby/examples/lib/apply_order_discount.rb

```diff
@@ -0,0 +1,44 @@
+class ApplyOrderDiscount
+  Request = Struct.new(:id)
+  Response = Struct.new(:discount, :has_discount?)
+
+  attr_reader :gateway
+
+  def initialize(gateway: Gateway.new)
+    @gateway = gateway
+  end
+
+  def apply(request)
+    order = gateway.find_order_by_id(request.id)
+
+    unless order.has_discount?
+      order.discount = discount_amount
+      gateway.save_order_discount(request.id, order)
+    end
+
+    Response.new(order.discount, order.has_discount?)
+  end
+
+  private
+
+  def discount_amount
+    10
+  end
+
+  class DiscountableOrder < Struct.new(:discount)
+    def has_discount?
+      discount > 0
+    end
+  end
+
+  class Gateway
```

Comment:
One of the points of Gateways is that they are reusable between multiple usecases. This reduces dev costs.

It's not completely about sharing/reusing either, it's about single responsibility. It causes this file to have multiple reasons to change. 

1. For database related reasons
2. For business logic related reasons

This may seem innocuous, but imagine this is a longly lived system with lots of code like this then you'll get into a situation where the lines between these responsibilities will become blurred. I think that by placing the class here you are effectively saying, as the committer of the code, "This class belongs to ApplyOrderDiscount, and I'm fine with you tightly coupling future calls to it"